pyhanko/sign/signers/cms_embedder.py

```diff
@@ -266,7 +266,10 @@ def apply(self, sig_annot, writer):
             stamp = self._appearance_stamp(
                 writer, BoxConstraints(width=w, height=h)
             )
-            sig_annot['/AP'] = stamp.as_appearances().as_pdf_object()
+            normalappearence = stamp.as_appearances()
+            if '/Matrix' in sig_annot:
```

Comment:
This isn't legal IIRC. Will double-check, but I'm pretty sure the matrix entry only goes in the appearance stream dictionary.

Will also review in the context of your remark in the "Caveats" section.

Reply:
So, I just had another look at this part:

> Currently this solution adds the transformation matrix to the signature field first and then copies it to the appearence of the signature. This could be prevented if the already added /N field is reused instead of overwritten.
Note: /N or rather /AP was added twice even without my change, once in fields.py at `ap_dict[pdf_name('/N')] = pdf_out.add_object(ap_stream)` and once in cms_embedder.py where my change adds the /Matrix field.


I checked the usage sites you mentioned. They both exist to serve slightly different code paths: one is for creating new signature fields (that may or may not be signed in the same session), the other is to add signature appearances as part of a signing process. So they're both necessary and independently useful, but it's indeed quite possible that some workflows hit both of these.

Reusing the XObject stream pointed to by `/N` when generating the signature appearance looks good at first sight, but I'm not convinced it's the way to go. If you're signing a previously-prepared signature field in a workflow with multiple signers you want to avoid updating existing stream objects in-place for forensics reasons that I won't go into.

Now, what would be OK to do is to _copy_ the `/Matrix` entry (if any) from the "old" `/N` into the new one. That's OK because it doesn't require modifying a stream object. If my reading of your code is correct, that should also achieve what you want without putting nonstandard keys into the annotation dictionary :).

Reply:
I honestly have no idea how that managed to be there and still work (or maybe i was hallucinating, who knows)
Is fixed now